tests/lib/rules/require-atomic-updates.js

```diff
@@ -0,0 +1,140 @@
+/**
+ * @fileoverview disallow assignments that can lead to race conditions due to usage of `await` or `yield`
+ * @author Teddy Katz
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const rule = require("../../../lib/rules/require-atomic-updates");
+const RuleTester = require("../../../lib/testers/rule-tester");
+
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 2018 } });
+
+const VARIABLE_ERROR = {
+    message: "Possible race condition: `foo` might be reassigned based on an outdated value of `foo`.",
+    type: "AssignmentExpression"
+};
+
+const STATIC_PROPERTY_ERROR = {
+    message: "Possible race condition: `foo.bar` might be reassigned based on an outdated value of `foo.bar`.",
+    type: "AssignmentExpression"
+};
+
+const COMPUTED_PROPERTY_ERROR = {
+    message: "Possible race condition: `foo[bar].baz` might be reassigned based on an outdated value of `foo[bar].baz`.",
+    type: "AssignmentExpression"
+};
+
+ruleTester.run("require-atomic-updates", rule, {
+
+    valid: [
+        "let foo; async function x() { foo += bar; }",
+        "let foo; async function x() { foo = foo + bar; }",
+        "let foo; async function x() { foo = await bar + foo; }",
+        "async function x() { let foo; foo += await bar; }",
+        "let foo; async function x() { foo = (await result)(foo); }",
```

Comment:
what if some stranger cases:
```js
let foo; async function x() { foo = (await result)(foo++); } // not sure
let foo; async function x() { foo = ((foo),(await result)(foo)); } // read the original, but valid

// invalid
let foo;
function bar(){foo++;}
async function x() { foo = bar() + await result }
```

Reply:
I'm not sure it's necessary to support these cases. In general, it's undecidable whether any given assignment creates a race condition, so the rule is necessarily just using a heuristic to detect races. As a result, there are always going to be some contrived cases where the rule has false positives/false negatives, so I'm not sure it's worth adding special logic for these cases unless the code looks like something that people would actually write.